lang/java/reef-utils/src/main/java/org/apache/reef/util/cache/CacheImpl.java

```diff
@@ -0,0 +1,116 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.reef.util.cache;
+
+import org.apache.reef.util.Optional;
+
+import javax.inject.Inject;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Implementation that supports expire-after-write.
+ * Entries that have expired are collected and invalidated on get calls.
+ * This obviates the need for a separate thread to invalidate expired entries, at
+ * the cost of some increase in get call latency.
+ * The invalidation sweep is only initiated after a check interval has passed,
+ * and at most one invalidation sweep is run at a time.
+ */
+public final class CacheImpl<K, V> implements Cache<K, V> {
+  private final ConcurrentMap<K, WrappedValue<V>> internalMap;
+  private final CurrentTime currentTime;
+  private final long timeoutMillis;
+  private final long expireCheckInterval;
+  private final AtomicBoolean expireInProgress;
+
+  private long expireCheckedTime;
+
+  /**
+   * Construct an expire-after-write cache
+   *
+   * @param currentTime   class that returns the current time for timeout purposes
+   * @param timeoutMillis a cache entry timeout after write
+   */
+  @Inject
+  public CacheImpl(final CurrentTime currentTime,
+                   final long timeoutMillis) {
+    this.internalMap = new ConcurrentHashMap<>();
+    this.currentTime = currentTime;
+    this.timeoutMillis = timeoutMillis;
+    this.expireCheckInterval = timeoutMillis / 2;
+    this.expireInProgress = new AtomicBoolean(false);
+
+    this.expireCheckedTime = currentTime.now();
+  }
+
+  @Override
+  public V get(final K key, final Callable<V> valueFetcher) throws ExecutionException {
```

Comment:
The inner-loop of `expireEntries` only runs once per `expireCheckInterval`. So the cost should be amortized across `get`s, but the `get` that runs it will indeed see increased latency if there are many entries.

I was thinking of pretty small clusters here. Maybe we need an (at least rough) performance target here?